File: frontend/src/test/utils/entries.ts
/**
 * Test data builders for entries + tags. Tiny defaults; per-test
 * overrides via the partial-overrides arg.
 */

import type {
  AppliedFilters,
  CalendarDay,
  CalendarResponse,
  EntryEnvelope,
  EntryListResponse,
  EntryResponse,
  EntryTagSummary,
  OnThisDayResponse,
  PassageEntriesResponse,
  ResolvedReference,
  TagListResponse,
  TagSummary,
} from "@/types/api";

let _entryId = 100;
let _tagId = 200;

export function nextEntryId(): number {
  _entryId += 1;
  return _entryId;
}

export function nextTagId(): number {
  _tagId += 1;
  return _tagId;
}

export function makeEntryTag(overrides: Partial<EntryTagSummary> = {}): EntryTagSummary {
  return {
    id: nextTagId(),
    name: "faith",
    ...overrides,
  };
}

export function makeEntry(overrides: Partial<EntryResponse> = {}): EntryResponse {
  const id = overrides.id ?? nextEntryId();
  return {
    id,
    title: null,
    display_title: "John 3:16",
    entry_date: "2026-05-27",
    scripture_ref: "John 3:16",
    translation_code: "BSB",
    scripture_text: "For God so loved the world.",
    observation: "",
    application: "",
    prayer: "",
    tags: [],
    created_at: "2026-05-27T00:00:00Z",
    updated_at: "2026-05-27T00:00:00Z",
    ...overrides,
  };
}

export function makeEntryEnvelope(
  overrides: Partial<EntryResponse> = {},
): EntryEnvelope {
  return { entry: makeEntry(overrides) };
}

export function makeEntryList(
  entries: EntryResponse[] = [makeEntry()],
  pagination: Partial<Omit<EntryListResponse, "entries" | "applied_filters">> = {},
  appliedFilters: Partial<AppliedFilters> = {},
): EntryListResponse {
  return {
    entries,
    total: pagination.total ?? entries.length,
    limit: pagination.limit ?? 20,
    offset: pagination.offset ?? 0,
    applied_filters: {
      q: appliedFilters.q ?? null,
      book: appliedFilters.book ?? null,
      tag: appliedFilters.tag ?? null,
      from_date: appliedFilters.from_date ?? null,
      to_date: appliedFilters.to_date ?? null,
    },
  };
}

export function makeTagSummary(overrides: Partial<TagSummary> = {}): TagSummary {
  return {
    id: nextTagId(),
    name: "faith",
    entry_count: 1,
    ...overrides,
  };
}

export function makeTagList(tags: TagSummary[] = []): TagListResponse {
  return { tags };
}

export function makeCalendarResponse(
  overrides: { year?: number; month?: number; days?: CalendarDay[] } = {},
): CalendarResponse {
  const year = overrides.year ?? 2026;
  const month = overrides.month ?? 5;
  const days = overrides.days ?? [];
  const total = days.reduce((sum, d) => sum + d.count, 0);
  return { year, month, days, total };
}

export function makeOnThisDayResponse(
  entries: EntryResponse[] = [],
  target_date = "2026-05-27",
): OnThisDayResponse {
  return { target_date, entries };
}

const DEFAULT_PASSAGE_REF: ResolvedReference = {
  canonical_string: "John 3",
  translation_code: "BSB",
  book: {
    name: "John",
    abbreviation: "John",
    order_index: 43,
    testament: "NT",
    chapter_count: 21,
  },
  chapter_number: 3,
  start_verse: 1,
  end_verse: 36,
};

export function makePassageEntriesResponse(
  entries: EntryResponse[] = [],
  reference: ResolvedReference = DEFAULT_PASSAGE_REF,
): PassageEntriesResponse {
  return { reference, count: entries.length, entries };
}

File: backend/soap_journal/schemas/entries.py
"""Pydantic request/response schemas for SOAP journal entries + tags."""

from __future__ import annotations

import unicodedata
from datetime import date, datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator

TAG_MAX_LEN = 50


class EntryTagSummary(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    name: str


class _EntryWritable(BaseModel):
    """Fields shared by create and update payloads.

    Both endpoints share this shape — PUT is replace-not-patch, so they
    are intentionally the same model.
    """

    title: str | None = None
    entry_date: date | None = None
    scripture_ref: str = Field(..., min_length=1, max_length=200)
    translation_code: str | None = None
    observation: str = ""
    application: str = ""
    prayer: str = ""
    tags: list[str] = Field(default_factory=list)

    @field_validator("title")
    @classmethod
    def _trim_title(cls, value: str | None) -> str | None:
        if value is None:
            return None
        trimmed = value.strip()
        if not trimmed:
            return None
        if len(trimmed) > 200:
            raise ValueError("title exceeds 200 characters")
        return trimmed

    @field_validator("tags")
    @classmethod
    def _validate_tags(cls, value: list[str]) -> list[str]:
        cleaned: list[str] = []
        for tag in value:
            if not isinstance(tag, str):
                raise ValueError("tags must be strings")
            stripped = tag.strip()
            if not stripped:
                raise ValueError("tag cannot be empty after trim")
            if len(stripped) > TAG_MAX_LEN:
                raise ValueError(f"tag exceeds {TAG_MAX_LEN} characters")
            if any(unicodedata.category(c).startswith("C") for c in stripped):
                raise ValueError("tag contains control characters")
            cleaned.append(stripped)
        return cleaned


class EntryCreateRequest(_EntryWritable):
    pass


class EntryUpdateRequest(_EntryWritable):
    pass


class EntryResponse(BaseModel):
    id: int
    title: str | None
    display_title: str
    entry_date: date
    scripture_ref: str
    translation_code: str
    scripture_text: str
    observation: str
    application: str
    prayer: str
    tags: list[EntryTagSummary]
    created_at: datetime
    updated_at: datetime


class EntryEnvelope(BaseModel):
    entry: EntryResponse


class AppliedFilters(BaseModel):
    q: str | None = None
    book: str | None = None
    tag: str | None = None
    from_date: date | None = None
    to_date: date | None = None


class EntryListResponse(BaseModel):
    entries: list[EntryResponse]
    total: int
    limit: int
    offset: int
    applied_filters: AppliedFilters


class CalendarDay(BaseModel):
    entry_date: date
    count: int


class CalendarResponse(BaseModel):
    year: int
    month: int
    days: list[CalendarDay]
    total: int


class OnThisDayResponse(BaseModel):
    target_date: date
    entries: list[EntryResponse]


# ---- tags ------------------------------------------------------------------


class TagSummary(BaseModel):
    id: int
    name: str
    entry_count: int


class TagListResponse(BaseModel):
    tags: list[TagSummary]


class TagAutocompleteResponse(BaseModel):
    tags: list[TagSummary]


# ---- list query ------------------------------------------------------------

EntryListOrder = Literal["newest", "oldest"]
